course_catalog/models.py

```diff
@@ -0,0 +1,24 @@
+from django.db import models
+from django.contrib.postgres.fields import JSONField
+
+
+class Course(models.Model):
+    title = models.CharField(max_length=256)
+    short_description = models.CharField(max_length=1024)
+    level = models.CharField(max_length=128)
+    semester = models.CharField(max_length=10)
+    language = models.CharField(max_length=128)
+    platform = models.CharField(max_length=128)
+    year = models.IntegerField()
+    id = models.CharField(max_length=128, unique=True)
+    full_description = models.TextField(null=True)
+    start_date = models.DateTimeField(null=True)
+    end_date = models.DateTimeField(null=True)
+    enrollment_start = models.DateTimeField(null=True)
+    enrollment_end = models.DateTimeField(null=True)
+    image = JSONField(null=True)
+    topics = JSONField()
+    prices = JSONField()
+    instructors = JSONField()
+    mode_types = JSONField()
+    raw_json = JSONField(null=True)
```

Comment:
What we're currently getting as JSON from the two data sources is really differently structured for instructors. If we really would like a consistent structure for instructors to use for the future, I believe we should create a separate `CourseInstructor` model now with defined fields. This means more work upfront now to define the model and write code that can deserialize from the two different structures into the same model, but it'll make future easier for tasks involving instructors. Should we create a `CourseInstructor` model now?

Reply:
Let's keep `raw_json` for now, like you suggested. The motivation is that if we'd like to display different metadata between courses in the sidebar modal that isn't currently being tracked in our Course model (e.g. the `department` for an OCW course), we will need to store the data somewhere.

Reply:
What is in these json fields ?  For example , what would be in "instructors" field ?

Reply:
I feel that models and explicitly defined fields for data that multiple sources will have in common, like instructor, would be more robust.  But I agree that this approach would require more time and effort, @Ferdi do you have a strong preference?

Reply:
As we discussed I suggest the main model should have common data.   So instructor field should have  First, last, middle, title.

Reply:
@Ferdi 
Here is an example of what the EDX staff(instructor) looks like:
```
"staff": [
    {
        "uuid": "4080e4a6-a477-4f8f-9ef9-be522fc6ea79",
        "salutation": null,
        "given_name": "Juichi",
        "family_name": "Yamagiwa",
        "bio": "As a world-renowned researcher and expert in the study of primatology, Juichi Yamagiwa served as Professor and Dean of Graduate School and Faculty of Science, followed by inauguration as the President of Kyoto University in 2014. Dr. Yamagiwa’s passion for fieldwork research frequently made him travel to the land of Africa, such as Gabon, DR Congo, where he discovered an abundance of new findings related to gorillas, through his unique viewpoint of evolution. His hobbies include lingala pop music and dance.",
        "slug": "juichi-yamagiwa",
        "position": {
            "title": "President",
            "organization_name": "Kyoto University",
            "organization_id": 25,
            "organization_override": null,
            "organization_marketing_url": null
        },
        "areas_of_expertise": [],
        "profile_image": {
            "medium": {
                "url": "https://prod-discovery.edx-cdn.org/media/people/profile_images/4080e4a6-a477-4f8f-9ef9-be522fc6ea79-2be36d827fd9.medium.png",
                "height": 110,
                "width": 110
            }
        },
        "works": [],
        "urls": {
            "twitter": null,
            "blog": null,
            "facebook": null
        },
        "urls_detailed": [],
        "email": null,
        "profile_image_url": "https://prod-discovery.edx-cdn.org/media/people/profile_images/4080e4a6-a477-4f8f-9ef9-be522fc6ea79-2be36d827fd9.png",
        "major_works": ""
    },
    {
        "uuid": "c3b48078-acff-4ced-ba43-8479ab5e7134",
        "salutation": null,
        "given_name": "Shun",
        "family_name": "Hongo",
        "bio": "Shun Hongo is a primate behavioral ecologist at the Primate Research Institute, Kyoto University. He has worked in a tropical forest in Gabon since 2009 for his research on society and ecology of mandrills and received his Ph. D. from Kyoto University in 2016. He assists in designing and conducting this course as a subject matter expert.",
        "slug": "shun-hongo",
        "position": {
            "title": "Researcher, Primate Research Institute",
            "organization_name": "Kyoto University",
            "organization_id": 25,
            "organization_override": null,
            "organization_marketing_url": null
        },
        "areas_of_expertise": [],
        "profile_image": {
            "medium": {
                "url": "https://prod-discovery.edx-cdn.org/media/people/profile_images/c3b48078-acff-4ced-ba43-8479ab5e7134-8d79173d7205.medium.png",
                "height": 110,
                "width": 110
            }
        },
        "works": [],
        "urls": {
            "twitter": null,
            "blog": null,
            "facebook": null
        },
        "urls_detailed": [],
        "email": null,
        "profile_image_url": "https://prod-discovery.edx-cdn.org/media/people/profile_images/c3b48078-acff-4ced-ba43-8479ab5e7134-8d79173d7205.png",
        "major_works": ""
    }
],
```
And below is OCW instructors example:
```
"instructors": [
    {
        "middle_initial": "",
        "first_name": "Michael",
        "last_name": "Short",
        "suffix": "",
        "title": "",
        "mit_id": "",
        "department": "",
        "directory_title": "",
        "uid": "d9ca5631c6936252866d63683c0c453e"
    },
    {
        "middle_initial": "",
        "first_name": "Jane",
        "last_name": "Kokernak",
        "suffix": "",
        "title": "",
        "mit_id": "",
        "department": "",
        "directory_title": "",
        "uid": "bb1e26b5f5c9c054ddae8a2988ad7b48"
    },
    {
        "middle_initial": "",
        "first_name": "Christine",
        "last_name": "Sherratt",
        "suffix": "",
        "title": "",
        "mit_id": "",
        "department": "",
        "directory_title": "",
        "uid": "a39f692061a70a105c25b15374c02c95"
    }
],
```

Reply:
Right, so the common fields seem to be first, middle , last , title.   That's what I would expect the db field to have , independent on the platform , right ?

Reply:
Yes for the most part, although I am not sure about middle as I couldn't find it in the EDX instructors

Reply:
ok, no middle name then. This is just as an example though. My point is that I thought we said the model will contain specific fields that are common . For the rest we have a json field for each platform to dump the respective json content as it comes from each platform.

Reply:
@Ferdi I was thinking more of a separate `Instructor` model with fields `first_name`, `last_name`, etc., along with a many-to-many relationship to the `Course` model (a course might have multiple instructors and an instructor might teach multiple courses).  But maybe that is getting too complicated for what we need the app to do.

Reply:
@mbertrand great, again, I'm not trying to prescribe a solution here. 

Related -- I think we should do the same for topics. A seperate model (course_tags)

Reply:
I agree. I'll let you and Matt decide.